libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefreshController.kt

```diff
@@ -21,30 +26,30 @@ import com.mapbox.navigation.utils.internal.logW
 internal class RouteRefreshController(
     private val routeRefreshOptions: RouteRefreshOptions,
     private val directionsSession: DirectionsSession,
-    private val tripSession: TripSession,
-    threadController: ThreadController,
+    private val currentLegIndexProvider: () -> Int,
+    private val threadController: ThreadController,
     private val routeDiffProvider: DirectionsRouteDiffProvider = DirectionsRouteDiffProvider(),
 ) {
 
+    private var currentRefreshRoutesJob: Job? = null
+
     internal companion object {
         internal const val LOG_CATEGORY = "RouteRefreshController"
     }
 
-    private val routerRefreshTimer = MapboxTimer(threadController).apply {
-        restartAfterMillis = routeRefreshOptions.intervalMillis
-    }
-
-    private var currentRequestId: Long? = null
-
     /**
      * If route refresh is enabled, attach a refresh poller.
      * Cancel old timers, and cancel pending requests.
      */
     fun restart(route: NavigationRoute, onRoutesRefreshed: (List<NavigationRoute>) -> Unit) {
         stop()
         if (route.routeOptions.enableRefresh() == true) {
-            routerRefreshTimer.startTimer {
-                refreshRoute(route, onRoutesRefreshed)
+            currentRefreshRoutesJob = threadController.getMainScopeAndRootJob().scope.launch {
+                refreshRoutes(route) { refreshedRoute ->
```

Comment:
The fact that this callback can fire multiple times, and for different reasons, is a little bit surprising. Also the name assumes refresh while it can also be cleanup.

What if we refactor this controller to have a flowable that's updated whenever a route is updated (either because of actual refresh or due to cleanup)? Then the client listens to the flowable all the time and occasionally calls `restart(route: NavigationRoute)` or `stop()` depending on what's happening. `stop()` might not even be needed if we organize it in coroutine scopes.

Reply:
Good point. I didn't like this aspect too and fixed it. I replaced callback by suspending function. The callback could be triggered many times but it didn't happen in real life. Whenever route refreshes, it triggers route update, which triggers restart of route refresh controller.